File: http-streams/src/main/java/http/PathMatching.java
package http;

import akka.http.javadsl.server.AllDirectives;
import akka.http.javadsl.server.PathMatcher1;
import akka.http.javadsl.server.PathMatchers;
import akka.http.javadsl.server.Route;

import static akka.http.javadsl.server.PathMatchers.*;

public class PathMatching extends AllDirectives {
  void seg() {
    //#segment
    Route route = path(PathMatchers.segment("hello"), () ->
      complete("Hello, World!"));
    //#segment
  }
  void staticimports() {
    //#staticimports
    Route route = path(segment("hello"), () ->
      complete("Hello, World!"));
    //#staticimports
  }
  void segments() {
    //#segments
    Route route = path(segment("hello").slash(segment()), name ->
      complete("Hello, " + name + "!"));
    //#segments
  }
  void advanced() {
    //#advanced
    PathMatcher1<Integer> m =
      PathMatchers
        .segment("foo")
        .slash("bar")
        .slash(
          segment("X").concat(integerSegment())
        )
        .slash(
          segment("edit").orElse(segment("create"))
        );
    //#advanced
  }
}

File: http-streams/src/main/scala/http/PathMatching.scala
package sc.http

import akka.http.scaladsl.server.{Directives, PathMatcher1}

object PathMatching extends Directives{
  private[http] def seg(): Unit = {
    //#segment
    val route = path("hello") {
      complete("Hello, World!")
    }
    //#segment
  }

  private[http] def segments(): Unit = {
    //#segments
    val route = path("hello" / Segment) { name =>
      complete("Hello, " + name + "!")
    }
    //#segments
  }

  private[http] def advanced(): Unit = {
    //#advanced
    val matcher: PathMatcher1[Option[Int]] =
      "foo" / "bar" / "X" ~ IntNumber.? / ("edit" | "create")
    //#advanced
  }
}
